File: ui/src/api-shared/types/ai.ts
/**
 * Domain contract for Bluz's AI assistant. Shared by `api-client` and
 * `api-server`, so it must stay free of side effects and of anything that only
 * exists on one side of the wire.
 *
 * Nothing here names a vendor: the backend model is expected to change, and a
 * provider swap must not ripple into route handlers, components, or types.
 */

/** Who produced a message in a conversation. */
export enum AiRole {
    /** Standing instructions. At most one, first in the list. */
    System = "system",
    /** Written by the human. */
    User = "user",
    /** Written by the model — may carry tool calls instead of text. */
    Assistant = "assistant",
    /** The result of running one tool the assistant asked for. */
    Tool = "tool",
}

/**
 * A tool invocation requested by the model. `arguments` stays a raw JSON
 * string, exactly as the model emitted it: it is not always valid JSON, and
 * the transcript replayed on the next turn must be byte-identical to what the
 * backend produced.
 */
export type AiToolCall = {
    id: string;
    name: string;
    arguments: string;
};

export type AiMessage = {
    role: AiRole;
    /** Empty on an assistant turn that only requested tools. */
    content: string;
    /** Assistant turns only. */
    toolCalls?: Array<AiToolCall>;
    /** Tool turns only — which call this answers. */
    toolCallId?: string;
    /** Tool turns only, for rendering the transcript. */
    name?: string;
};

/**
 * Whether running a tool changes stored data. Read tools run unattended;
 * write tools stop the turn and wait for the human to approve the exact call.
 */
export enum AiToolKind {
    Read = "read",
    Write = "write",
    /**
     * Not a data operation at all: the tool's whole effect is to put a
     * question to the human and stop the turn. Answered by the client, never
     * executed on the server.
     */
    Prompt = "prompt",
}

/**
 * How much damage running a tool can do. Drives the colour, the wording and
 * the extra confirmation on the approval card — a reversible edit and an
 * irreversible delete must not look alike.
 */
export enum AiToolDanger {
    /** Reads only, or a write that loses nothing. */
    Safe = "safe",
    /** Changes stored data; recoverable through history. */
    Caution = "caution",
    /** Destroys or overwrites data at scale. Needs an explicit re-confirm. */
    Destructive = "destructive",
}

/** A tool as advertised to the browser (for rendering, not for calling). */
export type AiToolSummary = {
    name: string;
    /** Friendly Hebrew label. The raw `name` is never shown to a user. */
    title: string;
    description: string;
    kind: AiToolKind;
    danger: AiToolDanger;
};

/** One selectable answer in an {@link AiStreamEventType.Choice} prompt. */
export type AiChoiceOption = {
    /** Sent back to the model verbatim. */
    value: string;
    /** Hebrew label on the button. */
    label: string;
    /** Optional one-line clarification under the label. */
    description?: string;
};

/** Request body of `POST /api/ai/chat`. */
export type ApiAiChatPayload = {
    /**
     * Full transcript, oldest first, excluding the system prompt — the server
     * owns that. Replay whatever the previous turn's `Done` event returned so
     * assistant tool calls and their results stay paired.
     */
    messages: Array<AiMessage>;
    /** The iteration the questions are about. Omitted means the current one. */
    iterationId?: string;
    /** Curriculum in view, if the user is on a Gantt screen. */
    curriculumId?: string;
    /**
     * Tool-call ids the human approved. A write tool runs only when its id is
     * listed here, so an approval covers one specific call and nothing else.
     */
    approvedToolCallIds?: Array<string>;
    /** Overrides the server default; normally unset so no vendor slug leaks. */
    model?: string;
};

export type AiUsage = {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
};

/**
 * Non-streaming result. The streaming route is the default path, but the
 * provider interface also answers in one shot for server-side callers (jobs,
 * CLI) that have nothing to stream to.
 */
export type AiChatResult = {
    content: string;
    toolCalls?: Array<AiToolCall>;
    model: string;
    usage?: AiUsage;
    /** Why generation ended — e.g. `stop`, `tool_calls`, `length`. */
    finishReason?: string;
};

export enum AiStreamEventType {
    /** A fragment of the assistant's visible answer. */
    Delta = "delta",
    /**
     * A fragment of the model's private reasoning. Rendered collapsed: it is
     * what makes an answer trustworthy, and noise the rest of the time.
     */
    Reasoning = "reasoning",
    /**
     * A fragment of a reasoning model's chain-of-thought, sent on a wire
     * channel separate from the visible answer. Shown collapsed by default.
     */
    ReasoningDelta = "reasoning_delta",
    /** The model is asking the human to pick between options. */
    Choice = "choice",
    /** A read tool started running. */
    ToolStart = "tool_start",
    /** A read tool finished; carries a short human-readable summary. */
    ToolResult = "tool_result",
    /** A write tool needs the human to approve it before it runs. */
    ToolProposal = "tool_proposal",
    /** Terminal success frame. */
    Done = "done",
    /** Terminal failure frame. */
    Error = "error",
}

/**
 * One frame of a streamed turn. A turn ends with exactly one `Done` **or**
 * one `Error`, never both.
 */
export type AiStreamEvent =
    | {
          type: AiStreamEventType.Choice;
          /** The `ask_user` call this answers; the client writes its result. */
          toolCallId: string;
          question: string;
          options: Array<AiChoiceOption>;
          /** Whether the human may type an answer instead of picking one. */
          allowFreeText: boolean;
      }
    | { type: AiStreamEventType.Delta; text: string }
    | {
          type: AiStreamEventType.Done;
          /**
           * The turn's new messages, to be appended to the transcript and
           * replayed on the next request. Includes assistant tool calls and
           * their tool results so an approved call resumes with its context.
           */
          messages: Array<AiMessage>;
          /** True when the turn stopped on a pending write approval. */
          awaitingApproval: boolean;
          model: string;
          usage?: AiUsage;
      }
    | {
          type: AiStreamEventType.Error;
          message: string;
          /**
           * The turn's produced messages so far, if any tool calls ran before
           * the failure. Optional: a pre-stream failure (auth, validation)
           * has none. When present the client should still append it to the
           * transcript before showing the error, so a retried turn does not
           * replay tool calls the server already executed.
           */
          messages?: Array<AiMessage>;
      }
    | { type: AiStreamEventType.Reasoning; text: string }
    | { type: AiStreamEventType.ReasoningDelta; text: string }
    | {
          type: AiStreamEventType.ToolProposal;
          toolCallId: string;
          name: string;
          /** Friendly Hebrew label for {@link name}. */
          title: string;
          danger: AiToolDanger;
          /** Parsed arguments, for showing the human what will change. */
          arguments: unknown;
          /** Hebrew, one line: what approving this will do. */
          summary: string;
          /** Hebrew bullets: the concrete consequences of approving. */
          impact?: Array<string>;
      }
    | {
          type: AiStreamEventType.ToolResult;
          toolCallId: string;
          name: string;
          title: string;
          summary: string;
          ok: boolean;
          /** Wall-clock duration of the call, for the timeline chip. */
          durationMs?: number;
          /** The envelope handed to the model, for the details panel. */
          detail?: unknown;
      }
    | {
          type: AiStreamEventType.ToolStart;
          toolCallId: string;
          name: string;
          title: string;
      };

/** Content type of the streaming chat route. */
export const AI_STREAM_CONTENT_TYPE = "text/event-stream";

/** Message cap per request — a guard against an unbounded prompt bill. */
export const AI_MAX_MESSAGES = 60;

/** Character cap on a single message's content. */
export const AI_MAX_MESSAGE_LENGTH = 16_000;

/**
 * Ceiling on tool round-trips inside one turn. A model that loops on a failing
 * tool would otherwise bill indefinitely.
 */
export const AI_MAX_TOOL_ITERATIONS = 8;

/**
 * Upstream completion cap, sent as `max_tokens` on every model call. Without
 * it a runaway or adversarial prompt has no ceiling on the bill for a single
 * response.
 */
export const AI_MAX_RESPONSE_TOKENS = 2_000;

/**
 * Character cap on one tool result before it is truncated. A tool that hands
 * the model a 200KB curriculum tree spends the whole context window on a
 * single call — and bills it again on every later turn of the conversation.
 */
export const AI_MAX_TOOL_RESULT_CHARS = 12_000;

/**
 * Starter prompts on the empty chat. The self-test keys a case by each one
 * (`SUGGESTED_PROMPT_CASES`), so a new chip without a case fails typecheck.
 */
export const AI_SUGGESTED_PROMPTS = [
    'מה יש בלו"ז השבוע?',
    "מי מבזר השבוע?",
    "תוסיף הפסקות בין שיעורים",
] as const;

export type AiSuggestedPrompt = (typeof AI_SUGGESTED_PROMPTS)[number];

File: cli/bluz_cli/commands/ai.py
"""
Name: ai.py
Purpose: Drive the in-app AI assistant from a terminal — capability report,
         a streaming chat turn (with the same write-tool approval gate the
         browser enforces), and the provider self-test.
         Mirrors ui/src/api-client/ai.ts.
Created: 2026-09-18
Author: Michael K. Steinberg
"""

from __future__ import annotations

import json
from typing import Any

import typer

from bluz_cli.commands._common import parse_json, show
from bluz_cli.context import state
from bluz_cli.output import console, success, warn

app = typer.Typer(help="The Bluz AI assistant.", no_args_is_help=True)

_BASE = "/api/ai"

# Frame types of the chat stream, mirroring AiStreamEventType in
# ui/src/api-shared/types/ai.ts. Spelled out here rather than imported (there
# is nothing to import from across the wire) and kept in one place so the
# renderer below never repeats a literal.
_DELTA = "delta"
_REASONING = {"reasoning", "reasoning_delta"}
_CHOICE = "choice"
_TOOL_START = "tool_start"
_TOOL_RESULT = "tool_result"
_TOOL_PROPOSAL = "tool_proposal"
_DONE = "done"
_ERROR = "error"


@app.command("tools")
def list_tools() -> None:
    """What the assistant can do, and whether AI is configured at all.

    Reports `enabled: false` on a deployment with no model key rather than
    failing — the same signal the UI uses to hide the launcher.
    """
    with state.client() as client:
        show(client.get(f"{_BASE}/tools"), title="Assistant")


@app.command("benchmark")
def benchmark() -> None:
    """Run the assistant self-test against a fabricated fixture.

    Throttled server-side to once an hour per user (a run costs real tokens),
    which comes back as an `AiRateLimitError`.
    """
    with state.client() as client:
        result = client.post(f"{_BASE}/benchmark")
    success("Benchmark finished")
    show(result, title="Benchmark")


def _proposal_line(event: dict[str, Any]) -> str:
    """One human-readable line describing a pending write-tool call."""
    arguments = event.get("arguments")
    rendered = (
        json.dumps(arguments, ensure_ascii=False) if arguments is not None else ""
    )
    return f"{event.get('title') or event.get('name')}: {event.get('summary', '')} {rendered}".strip()


def _render_stream(events, *, as_json: bool) -> dict[str, Any] | None:
    """Consume one turn's frames, printing as they arrive.

    Returns the terminal `done` frame (or None when the turn ended in an
    error), so the caller can decide whether an approval round is needed.
    """
    done: dict[str, Any] | None = None
    pending: list[dict[str, Any]] = []
    wrote_text = False

    for event in events:
        kind = event.get("type")
        if as_json:
            print(json.dumps(event, ensure_ascii=False, default=str))
        elif kind == _DELTA:
            console.print(event.get("text", ""), end="")
            wrote_text = True
        elif kind in _REASONING:
            # Reasoning is collapsed in the browser; dim it here for the same
            # reason — it is context, not the answer.
            console.print(f"[dim]{event.get('text', '')}[/dim]", end="")
        elif kind == _TOOL_START:
            console.print(f"\n[cyan]→ {event.get('title') or event.get('name')}[/cyan]")
        elif kind == _TOOL_RESULT:
            colour = "green" if event.get("ok") else "red"
            console.print(
                f"[{colour}]← {event.get('title') or event.get('name')}: "
                f"{event.get('summary', '')}[/{colour}]"
            )
        elif kind == _CHOICE:
            options = ", ".join(
                str(option.get("label")) for option in event.get("options", [])
            )
            console.print(f"\n[yellow]? {event.get('question')}[/yellow] ({options})")
        elif kind == _TOOL_PROPOSAL:
            console.print(f"\n[yellow]⚠ {_proposal_line(event)}[/yellow]")
        elif kind == _ERROR:
            if wrote_text:
                console.print()
            console.print(f"[red]✗ {event.get('message', 'Stream failed')}[/red]")

        if kind == _TOOL_PROPOSAL:
            pending.append(event)
        elif kind == _DONE:
            done = event
        elif kind == _ERROR:
            done = None

    if not as_json and wrote_text:
        console.print()
    if done is not None:
        done["_pending"] = pending
    return done


@app.command("chat")
def chat(
    prompt: str = typer.Argument(..., help="What to ask the assistant."),
    iteration: str = typer.Option(
        None, "--iteration", "--it", help="Iteration the question is about."
    ),
    curriculum_id: str = typer.Option(
        None, "--curriculum-id", help="Curriculum in view, for gantt questions."
    ),
    model: str = typer.Option(None, "--model", help="Override the server's model."),
    messages: str = typer.Option(
        None,
        "--messages",
        help="JSON array continuing a previous transcript (a `done` frame's messages).",
    ),
    approve: list[str] = typer.Option(
        None,
        "--approve",
        help="Tool-call id to approve, repeatable. An approval covers that one call.",
    ),
    yes: bool = typer.Option(
        False,
        "--yes",
        "-y",
        help="Approve every write tool the turn proposes, and resume automatically.",
    ),
) -> None:
    """Ask the assistant one question and stream the answer.

    Write tools stop the turn and wait for a human, exactly as in the browser.
    Re-run with `--approve <toolCallId>` (plus `--messages` carrying the
    transcript the `done` frame returned) to let one through, or pass `--yes`
    to approve and resume in a single invocation.
    """
    transcript = parse_json(messages, what="--messages") or []
    if not isinstance(transcript, list):
        raise typer.BadParameter("--messages must be a JSON array of messages.")

    payload: dict[str, Any] = {
        "messages": [*transcript, {"role": "user", "content": prompt}],
        "iterationId": iteration,
        "curriculumId": curriculum_id,
        "model": model,
        "approvedToolCallIds": list(approve or []),
    }
    payload = {key: value for key, value in payload.items() if value not in (None, [])}

    with state.client() as client:
        done = _render_stream(
            client.stream_sse(f"{_BASE}/chat", json=payload),
            as_json=state.as_json,
        )

        # A turn that stopped on a pending write is only half an answer. With
        # --yes, resume it immediately by replaying the transcript with the
        # proposed call ids approved; without it, say what is waiting and how.
        while done is not None and done.get("awaitingApproval"):
            pending = done.get("_pending") or []
            ids = [
                event.get("toolCallId") for event in pending if event.get("toolCallId")
            ]
            if not yes:
                warn(
                    "Turn is waiting for approval of: "
                    + ", ".join(_proposal_line(event) for event in pending)
                )
                warn(
                    "Re-run with --approve "
                    + " --approve ".join(str(i) for i in ids)
                    + " (and --messages to carry the transcript), or pass --yes."
                )
                break
            if not ids:
                break
            # The `done` frame carries the turn's *new* messages, to be
            # appended to what was sent — replacing the transcript with them
            # would drop the question the approval is about.
            payload = {
                **payload,
                "messages": [*payload["messages"], *done.get("messages", [])],
                "approvedToolCallIds": ids,
            }
            done = _render_stream(
                client.stream_sse(f"{_BASE}/chat", json=payload),
                as_json=state.as_json,
            )
